lib/acts_as_list/active_record/acts/list.rb

```diff
@@ -389,7 +392,28 @@ def shuffle_positions_on_intermediate_items(old_position, new_position, avoid_id
 
         def insert_at_position(position, raise_exception_if_save_fails=false)
           raise ArgumentError.new("position cannot be lower than top") if position < acts_as_list_top
-          return set_list_position(position, raise_exception_if_save_fails) if new_record?
+          if invalid?
+            # If you violate a uniqueness constraint on
+            # position_column, you get a postgres error,
+            # and the transaction rolls back.
+            # You can avoid this with a uniqueness validation (raising
+            # RecordNotValid) which you can catch and continue from.
+            save! if raise_exception_if_save_fails
+            return false
+          end
+          return set_list_position(position, raise_exception_if_save_fails) if new_record? && not_in_list?
+
+          # Cannot lock a record with unsaved changes, so save it first.
+          # If the changes are only to the position column, reset them, since we are overriding them.
+          self[position_column] = position_before_save if position_before_save_changed?
```

Comment:
The point of this call is to avoid a save that isn't required, so I don't want `set_list_position`. I'll update it to `write_attribute`.

Reply:
Setting the attribute is to avoid a save call :) I've changed this and several other instances of `self[position_column] =` to `write_attribute` though, thanks.